### PR TITLE
feat(api): add fft

### PR DESCRIPTION
**Pull request by minghuaw**

### CODE REVIEW

File: src/stream.rs
Comment (dcvz) [inline src/stream.rs:87]: We could name this `default_stream_for_device()`?

Comment (minghuaw) [reply src/stream.rs:87]: Sounds good. I will make that change in this PR

Comment (minghuaw) [reply src/stream.rs]: Renamed to `default_stream_on_device`

**Comment by dcvz** [comment]
@minghuaw I wonder if you can leverage your ported `fftn`, `ifftn` , `irfftn` and `rfftn` for  `fft2`, `ifft2`, `irfft2` and `rfft2`  - might make validations a tiny bit simpler eventhough you've already tried to generalize checks a lot 
internally it does something similar as well and it might allow you to have less inners

**Comment by dcvz** [comment]
It's all built on top of the "n" variants with default axis/axes: https://github.com/ml-explore/mlx/blob/d0dbfe0b97b95bc29009e7534f7ff734fb679801/mlx/fft.h#L120

File: src/fft/fftn.rs
Comment (dcvz) [inline src/fft/fftn.rs:16]:
```suggestion
///   with zeros to match `s`. The default value is `a.shape[axis]`.
```

File: src/fft/fftn.rs
Comment (dcvz) [inline src/fft/fftn.rs:37]: for some you call this `n` and for some `s`

File: src/fft/fftn.rs
Comment (dcvz) [inline src/fft/fftn.rs:81]:
```suggestion
///  with zeros to match `s`. The default value is the sizes of `a` along `axes`.
```

File: src/fft/fftn.rs
Comment (dcvz) [inline src/fft/fftn.rs:100]:
```suggestion
/// with zeros to match `s`. The default value is the sizes of `a` along `axes`.
```

File: src/fft/fftn.rs
Comment (dcvz) [inline src/fft/fftn.rs:119]:
```suggestion
/// with zeros to match `s`. The default value is the sizes of `a` along `axes`.
```

File: src/fft/fftn.rs
Comment (dcvz) [inline src/fft/fftn.rs:288]:
```suggestion
/// with zeros to match `s`. The default value is the sizes of `a` along `axes`.
```

File: src/fft/fftn.rs
Comment (dcvz) [inline src/fft/fftn.rs:307]:
```suggestion
/// with zeros to match `s`. The default value is the sizes of `a` along `axes`.
```

File: src/fft/fftn.rs
Comment (dcvz) [inline src/fft/fftn.rs:326]:
```suggestion
/// with zeros to match `s`. The default value is the sizes of `a` along `axes`.
```

File: src/fft/mod.rs
Comment (dcvz) [inline src/fft/mod.rs:184]: minor: I'd love to move these to an `fft/utils.rs` just to make this module easy to digest for those looking.

File: src/fft/fftn.rs
Comment (dcvz) [inline src/fft/fftn.rs:423]: I like to add tests for known failures -- but if you think its enough to be covered by the tests in the utils you're using to validate, then I leave it up to you.

**Review by dcvz — APPROVED** [review]
LGTM 👍 one suggestion for the validations and some inconsistencies with docs (i didn't list them all) but otherwise this LGTM. Approving so you can merge when you've resolved comments.

File: src/error.rs
Comment (minghuaw) [reply src/error.rs:63]: Didn't do this because the c++ exception's not doing it

File: src/fft/fftn.rs
Comment (minghuaw) [reply src/fft/fftn.rs:16]: This follows the python api. 1D fft uses `n`, n-D uses `s`

File: src/fft/fftn.rs
Comment (minghuaw) [reply src/fft/fftn.rs:37]: This follows the python api. 1D fft uses n, n-D uses s

File: src/fft/fftn.rs
Comment (minghuaw) [reply src/fft/fftn.rs:81]: This follows the python api. 1D fft uses n, n-D uses s

File: src/fft/fftn.rs
Comment (minghuaw) [reply src/fft/fftn.rs:100]: This follows the python api. 1D fft uses n, n-D uses s

File: src/fft/fftn.rs
Comment (minghuaw) [reply src/fft/fftn.rs:119]: This follows the python api. 1D fft uses n, n-D uses s

File: src/fft/fftn.rs
Comment (minghuaw) [reply src/fft/fftn.rs:288]: This follows the python api. 1D fft uses n, n-D uses s

File: src/fft/fftn.rs
Comment (dcvz) [reply src/fft/fftn.rs:100]: In this case though, you used `s` as a name but then used `n` here. so there's a mismatch

Comment (minghuaw) [reply src/fft/fftn.rs:100]: Yeah, didn't realize that

File: src/fft/fftn.rs
Comment (minghuaw) [reply src/fft/fftn.rs:423]: I think this would suffice

**Pull request merged**
